### PR TITLE
Support subquery in new plan framework

### PR DESCRIPTION
**Pull request by shenli**
@coocood @qiuyesuifeng

### CODE REVIEW

File: ast/expressions.go [hunk 1]
```diff
@@ -242,11 +261,25 @@ func (n *SubqueryExpr) Accept(v Visitor) (Node, bool) {
...
+
+	if n.SubQuery != nil {
+		t, ok := n.SubQuery.Accept(v)
+		if !ok {
+			return n, false
+		}
+		sq, ok := t.(SubQuery)
+		if ok {
+			n.SubQuery = sq
+		}
```
Comment (coocood) [inline ast/expressions.go]: should return error if `t` is not `SubQuery`

File: ast/expressions.go [hunk 2]
```diff
@@ -228,11 +229,29 @@ func (n *CaseExpr) Accept(v Visitor) (Node, bool) {
 	return v.Leave(n)
 }
 
+// SubQuery represents a subquery interface.
+// This interface is used in plan/executor/evaluator
```
Comment (shenli) [inline ast/expressions.go]: .

File: executor/executor.go [hunk 3]
```diff
@@ -1075,5 +1075,36 @@ func (e *AggregateExec) Close() error {
...
 		return e.Src.Close()
 	}
+	for _, af := range e.AggFuncs {
+		af.Clear()
+	}
 	return nil
 }
+
+// SubqueryExec represents a subquery executor.
+type SubqueryExec struct {
```
Comment (shenli) [inline executor/executor.go]: Delete this.

File: executor/subquery.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/optimizer/plan"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// SubQuery is an exprNode with a plan.
+type subquery struct {
+	ast.ExprNode
```
Comment (coocood) [inline executor/subquery.go]: `ast.ExprNode` is not used.

File: executor/subquery.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		}
+		if len(row.Data) == 1 {
+			rows = append(rows, row.Data[0])
+		} else {
+			rows = append(rows, row.Data)
+		}
+		if rowCount > 0 {
+			rowCount--
+		}
+	}
```
Comment (coocood) [inline executor/subquery.go]: Should close `e`.

File: optimizer/evaluator/evaluator.go [hunk 6]
```diff
@@ -78,12 +78,24 @@ func boolToInt64(v bool) int64 {
 
 // Evaluator is an ast Visitor that evaluates an expression.
 type Evaluator struct {
-	ctx context.Context
-	err error
+	ctx          context.Context
+	err          error
+	multipleRows bool
```
Comment (coocood) [inline optimizer/evaluator/evaluator.go]: Will this be used in inner subquery in a subquery?

File: optimizer/evaluator/evaluator.go [hunk 7]
```diff
@@ -211,24 +228,147 @@ func (e *Evaluator) caseExpr(v *ast.CaseExpr) bool {
...
+
+func (e *Evaluator) subquery(v *ast.SubqueryExpr) bool {
+	if v.SubQuery != nil {
+		v.SetValue(v.SubQuery.GetValue())
+	}
+	v.Evaluated = true
+	return true
+}
+
+func (e *Evaluator) subqueryEval(v ast.SubQuery) bool {
```
Comment (coocood) [inline optimizer/evaluator/evaluator.go]: Add more comments.

File: optimizer/plan/planbuilder.go [hunk 8]
```diff
@@ -532,3 +549,30 @@ func splitWhere(where ast.ExprNode) []ast.ExprNode {
...
+
+// SubQueryBuilder is the interface for building SubQuery executor.
+type SubQueryBuilder interface {
+	Build(p Plan) ast.SubQuery
+}
+
+type subqueryVisitor struct {
+	subqueries []ast.SubQuery
+	builder    *planBuilder
+	sb         SubQueryBuilder
```
Comment (coocood) [inline optimizer/plan/planbuilder.go]: sb is already is builder.

File: optimizer/plan/subquery.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/infoschema"
+)
+
+// SubQuery is an exprNode with a plan.
+type SubQuery struct {
```
Comment (coocood) [inline optimizer/plan/subquery.go]: This can be deleted.

File: optimizer/evaluator/evaluator.go (hunk 6 above)
Comment (shenli) [inline optimizer/evaluator/evaluator.go]: No. For nested subquery, there will be a new evaluator.

**Comment by coocood** [comment]
LGTM

File: ast/expressions.go [hunk 10]
```diff
@@ -228,11 +229,29 @@ func (n *CaseExpr) Accept(v Visitor) (Node, bool) {
...
 }
 
+// SubQuery represents a subquery interface.
+// This interface is implemented in executor and used in plan/evaluator.
+type SubQuery interface {
+	ExprNode
+
+	// EvalRows executes the subquery and returns the multi rows with rowCount.
+	// rowCount < 0 means no limit.
+	// If the ColumnCount is 1, we will return a column result like {1, 2, 3},
```
Comment (qiuyesuifeng) [inline ast/expressions.go]: ColumnCount -> rowCount?

File: ast/expressions.go [hunk 11]
```diff
@@ -228,11 +229,29 @@ func (n *CaseExpr) Accept(v Visitor) (Node, bool) {
...
+	ColumnCount() (int, error)
+}
+
 // SubqueryExpr represents a subquery.
 type SubqueryExpr struct {
 	exprNode
 	// Query is the query SelectNode.
-	Query ResultSetNode
+	Query           ResultSetNode
+	SubQuery        SubQuery
```
Comment (qiuyesuifeng) [inline ast/expressions.go]: Seems we should use a new name, SubQuery and SubqueryExpr are a little confused.

File: executor/compiler.go [hunk 12]
```diff
@@ -35,14 +37,19 @@ type Compiler struct {
...
 		is := sessionctx.GetDomain(ctx).InfoSchema()
 		if err := optimizer.Preprocess(node, is, ctx); err != nil {
 			return nil, errors.Trace(err)
 		}
-		p, err := optimizer.Optimize(ctx, node)
+		// Validate should be after NameResolve.
+		if err := optimizer.Validate(node, false); err != nil {
+			return nil, errors.Trace(err)
+		}
+		sb := &subqueryBuilder{
```
Comment (qiuyesuifeng) [inline executor/compiler.go]: How about using NewSubQueryBuilder function?

File: executor/subquery.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+		return nil, errors.Trace(b.err)
+	}
+	defer e.Close()
+	if len(e.Fields()) == 0 {
+		// No result fields means no Recordset.
+		for {
+			row, err := e.Next()
+			if err != nil || row == nil {
+				return nil, errors.Trace(err)
+			}
```
Comment (qiuyesuifeng) [inline executor/subquery.go]: We should check err first, then row value.

File: executor/subquery.go [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+	b := newExecutorBuilder(ctx, sq.is)
+	plan.Refine(sq.plan)
+	e := b.build(sq.plan)
+	if b.err != nil {
+		return nil, errors.Trace(b.err)
+	}
+	defer e.Close()
+	if len(e.Fields()) == 0 {
+		// No result fields means no Recordset.
+		for {
```
Comment (qiuyesuifeng) [inline executor/subquery.go]: Seems row has been discarded, no need to using break?

File: optimizer/evaluator/evaluator.go [hunk 15]
```diff
@@ -211,24 +228,150 @@ func (e *Evaluator) caseExpr(v *ast.CaseExpr) bool {
...
-func (e *Evaluator) columnName(v *ast.ColumnNameExpr) bool {
-	v.SetValue(v.Refer.Expr.GetValue())
+func (e *Evaluator) compareSubquery(cs *ast.CompareSubqueryExpr) bool {
+	lv := cs.L.GetValue()
+	if lv == nil {
+		cs.SetValue(nil)
+		return true
+	}
+	x, err := e.checkResult(cs, lv, cs.R.GetValue().([]interface{}))
+	if err != nil {
```
Comment (qiuyesuifeng) [inline optimizer/evaluator/evaluator.go]: Should we log this error?

File: optimizer/evaluator/evaluator.go [hunk 16]
```diff
@@ -265,29 +412,27 @@ func (e *Evaluator) patternIn(n *ast.PatternInExpr) bool {
...
-		if types.IsNil(v.GetValue()) {
-			hasNull = true
-			continue
+	if n.Sel == nil {
+		values := make([]interface{}, 0, len(n.List))
+		for _, ei := range n.List {
+			values = append(values, ei.GetValue())
 		}
-		r, err := types.Compare(n.Expr.GetValue(), v.GetValue())
+		x, err := e.checkInList(n.Not, lhs, values)
```
Comment (qiuyesuifeng) [inline optimizer/evaluator/evaluator.go]: Should we set `e.err = errors.Trace(err)` like old code do?

File: optimizer/evaluator/evaluator_binop.go [hunk 17]
```diff
@@ -154,7 +154,7 @@ func (e *Evaluator) handleComparisonOp(o *ast.BinaryOperationExpr) bool {
 		// for <=>, if a and b are both nil, return true.
 		// if a or b is nil, return false.
 		if o.Op == opcode.NullEQ {
-			if types.IsNil(a) || types.IsNil(b) {
+			if types.IsNil(a) && types.IsNil(b) {
```
Comment (qiuyesuifeng) [inline optimizer/evaluator/evaluator_binop.go]: Is there a bug in old code?

File: optimizer/plan/planbuilder.go [hunk 18]
```diff
@@ -154,12 +157,25 @@ func (b *planBuilder) extractSelectAgg(sel *ast.SelectStmt) []*ast.AggregateFunc
 	return extractor.AggFuncs
 }
 
+func (b *planBuilder) buildSubquery(n ast.Node) {
+	sv := &subqueryVisitor{
+		builder: b,
+	}
+	_, ok := n.Accept(sv)
+	if !ok {
+		fmt.Println("Extract subquery error")
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder.go]: Using log for tracing.

File: optimizer/plan/planbuilder_join.go [hunk 19]
```diff
@@ -584,7 +584,9 @@ func (b *planBuilder) buildJoin(sel *ast.SelectStmt) Plan {
 	whereConditions := splitWhere(sel.Where)
 	for _, whereCond := range whereConditions {
 		if !path.attachCondition(whereCond, nil) {
-			log.Errorf("Failed to attach where condition.")
+			// TODO: Find a better way to handle this condition.
+			path.conditions = append(path.conditions, whereCond)
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder_join.go]: Could u show an example for this modification?

File: optimizer/evaluator/evaluator_binop.go (hunk 17 above)
Comment (shenli) [inline optimizer/evaluator/evaluator_binop.go]: Yes

File: optimizer/plan/planbuilder_join.go (hunk 19 above)
Comment (shenli) [inline optimizer/plan/planbuilder_join.go]: "select c1 from t1 where c1 in (select c2 from t2 where t2.c1=t1.c1)"
This is a temp solution. cooocood will fix it after this pr.

File: executor/subquery.go (hunk 14 above)
Comment (shenli) [inline executor/subquery.go]: There is not result. So we do not need the row.

File: ast/expressions.go (hunk 10 above)
Comment (shenli) [inline ast/expressions.go]: No. It is just ColumnCount. {{1}, {2}, {3}} will be {1, 2, 3}

**Comment by shenli** [comment]
@qiuyesuifeng PTAL

**Pull request merged**